File: frontend/lib/platformDetect.ts
export type Platform = "youtube" | "instagram" | "facebook" | "tiktok";

const PATTERNS: Record<Platform, RegExp[]> = {
  youtube: [
    /(?:https?:\/\/)?(?:www\.)?youtube\.com\/watch\?v=/i,
    /(?:https?:\/\/)?(?:www\.)?youtu\.be\//i,
    /(?:https?:\/\/)?(?:www\.)?youtube\.com\/shorts\//i,
  ],
  instagram: [
    /(?:https?:\/\/)?(?:www\.)?instagram\.com\/(?:p|reel|tv)\//i,
  ],
  facebook: [
    /(?:https?:\/\/)?(?:www\.)?facebook\.com\/.+\/videos\//i,
    /(?:https?:\/\/)?(?:www\.)?fb\.watch\//i,
    /(?:https?:\/\/)?(?:www\.)?facebook\.com\/watch/i,
    /(?:https?:\/\/)?(?:www\.)?facebook\.com\/reel\//i,
    /(?:https?:\/\/)?(?:www\.)?facebook\.com\/share\/v\//i,
    /(?:https?:\/\/)?(?:www\.)?facebook\.com\/share\/r\//i,
    /(?:https?:\/\/)?m\.facebook\.com\//i,
  ],
  tiktok: [
    /(?:https?:\/\/)?(?:www\.)?tiktok\.com\/@[^/]+\/video\//i,
    /(?:https?:\/\/)?vm\.tiktok\.com\//i,
    /(?:https?:\/\/)?vt\.tiktok\.com\//i,
    /(?:https?:\/\/)?m\.tiktok\.com\//i,
  ],
};

export function detectPlatform(url: string): Platform | null {
  for (const [platform, patterns] of Object.entries(PATTERNS) as [Platform, RegExp[]][]) {
    if (patterns.some((p) => p.test(url))) {
      return platform;
    }
  }
  return null;
}

export function isValidUrl(url: string): boolean {
  return /^https?:\/\//i.test(url.trim());
}

File: backend/utils/platform_detect.py
from __future__ import annotations
import re

PATTERNS = {
    "youtube": [
        r"(?:https?://)?(?:www\.)?youtube\.com/watch\?v=",
        r"(?:https?://)?(?:www\.)?youtu\.be/",
        r"(?:https?://)?(?:www\.)?youtube\.com/shorts/",
    ],
    "instagram": [
        r"(?:https?://)?(?:www\.)?instagram\.com/(?:p|reel|tv)/",
    ],
    "facebook": [
        r"(?:https?://)?(?:www\.)?facebook\.com/.+/videos/",
        r"(?:https?://)?(?:www\.)?fb\.watch/",
        r"(?:https?://)?(?:www\.)?facebook\.com/watch",
        r"(?:https?://)?(?:www\.)?facebook\.com/reel/",
        r"(?:https?://)?(?:www\.)?facebook\.com/share/v/",
        r"(?:https?://)?(?:www\.)?facebook\.com/share/r/",
        r"(?:https?://)?m\.facebook\.com/",
    ],
    "tiktok": [
        r"(?:https?://)?(?:www\.)?tiktok\.com/@[^/]+/video/",
        r"(?:https?://)?vm\.tiktok\.com/",
        r"(?:https?://)?vt\.tiktok\.com/",
        r"(?:https?://)?m\.tiktok\.com/",
    ],
}


def detect_platform(url: str) -> str | None:
    for platform, patterns in PATTERNS.items():
        for pattern in patterns:
            if re.search(pattern, url, re.IGNORECASE):
                return platform
    return None


def is_valid_url(url: str) -> bool:
    return bool(re.match(r"https?://", url.strip()))
